### ISSUE TITLE
Crash with large grammars and parse trees

### ISSUE THREAD

**Issue by kaby76** [zspitz/ANTLR4ParseTreeVisualizer#40]
Attached is a zip file of a solution/project containing a Java parser example. Displaying the tree with ANTLR4ParseTreeVisualizer crashes, presumably because it is a large parse tree, on release 0.1.0.63. It's good to be able to visualize non-trivial large trees, especially for nodes deep in the tree in a tree visitor. GetText() of a node doesn't work well, and I need to visualize where in the tree I am, e.g., with a variableDeclarator in what block.
[AntlrConsole3.zip](https://github.com/zspitz/ANTLR4ParseTreeVisualizer/files/3801883/AntlrConsole3.zip)

**Comment by zspitz** [comment]
> `GetText()` of a node doesn't work well

Is this related to #37?

**Comment by kaby76** [comment]
Yes, Antlr's runtime RuleContext.GetText() doesn't fill in white space between tokens. https://github.com/antlr/antlr4/blob/master/runtime/CSharp/runtime/CSharp/Antlr4.Runtime/RuleContext.cs#L173 . It's annoying, but one can write an extension method that works around it. The bug is that the visualizer crashes in the attached program after the parse, trying to display "tree". Adding in a listener for variableDeclarator and breaking there instead, I can see the parse tree for "context". Note, I could get the parse tree pane to show the node for version 0.1.0.65, but not for 0.1.0.66--"No tree nodes. Choose the lexer class, the parser class, and the parse method from the settings." I could select a parser and lexer but not a parser rule. I'm not clear why this would not be needed--you have the tree node when I clicked on the magnifier icon to call the visualizer.

**Comment by zspitz** [comment]
@kaby76 

Aside from the `GetText` issue, I've fixed the visibility binding logic, and the treeview and tokens display properly. I've pushed a commit, and it's in 
https://github.com/zspitz/ANTLR4ParseTreeVisualizer/releases/tag/0.1.0.68.

> I'm not clear why this would not be needed--you have the tree node when I clicked on the magnifier icon to call the visualizer.

You can now run the visualizer against a string (or a token stream); but in order to do so, you need to choose which lexer, parser, and parse method to use. If there is only one lexer defined outside of the Antlr.Runtime or Antlr.Runtime.Standard DLLs, that is set as the default. Same goes for the parser. The visibility was tied to whether there was a selected lexer and parser and parser method, which is wrong if the the visualizer is run against a parse tree. As I said, it should be working now.

I didn't see any sign of a crash in the project you uploaded. Perhaps a specifically deeply nested node?

**Comment by zspitz** [comment]
> and I need to visualize where in the tree I am, e.g., with a variableDeclarator in what block.

Could you elaborate on what you mean here?

**Comment by kaby76** [comment]
Wonderful! 0.1.0.68 fixes the crash and the parse tree pane problem. Now, it would be even better when I want to visualize a particular node in a listener method to display the tree in one of two ways, selectable with some sort of UI: (1) the whole tree (from root), with the focus set to the node I asked to display; (2) display the partial tree, starting from the internal node and into the sub-tree as you currently do. You could add an item to the right-click context menu which you have already ("Scope to this;  Scope to this in new window/; ...") to toggle between either of these views of the parse tree. This will really help to visualize where I am in the parse tree.

Eventually, it would be even better to highlight the code in an ITextView back in VS for the node I selected. By default, Antlr4.Runtime.Tree.TerminalNodeImpl.Symbol.InputStree.SourceName is null, but if you assume that the programmer here sets AntlrInputStream.name to the full path of the file after creating the AntlrInputStream, then your code could then open a fresh ITextView for the file. You'd also need to change the visualizer dialog box to be non-modal so I can go and look at the code in a VS ITextView.

One other request--double click on the Parent in Properties to set the Parse Tree View to a new node. You might then want to add a "back" function to navigate back to where you were before double-clicking on the Parent property.

Thanks so much! Looking good.

**Comment by zspitz** [comment]
RE: GetText -- I've committed fix for #37 with an [extension method](https://github.com/zspitz/ANTLR4ParseTreeVisualizer/blob/e8bc5613626310eb4753daa21f3abc4137f94b48/Shared/Util/Extensions/IParseTree.cs#L26) that serves as a replacement for GetText but fills in the blanks.

RE viewing the entire tree from a single node, I've opened #34 

Navigating to Parent node from Properties is in #44 .

See also #45

**Comment by zspitz** [comment]
@kaby76 Just a last question: could you upload an even larger Java source file, say 5000 characters? I want to see how the source textbox handles scrolling in various directions, and does it keep the selection in the right place? Thanks.

**Comment by kaby76** [comment]
I tried the visualizer with input https://github.com/kaby76/AntlrVSIX/blob/master/test/FormatterTest.java. I got an interesting crash:

`
   at Microsoft.VisualStudio.DebuggerVisualizers.Impl.VisualizerObjectProvider.TransferDataToDebuggeeSideVisualizer(Stream outgoingData)
   at Microsoft.VisualStudio.DebuggerVisualizers.Impl.VisualizerObjectProvider.Microsoft.VisualStudio.DebuggerVisualizers.IVisualizerObjectProvider.TransferObject(Object outgoingObject)
   at ParseTreeVisualizer.VisualizerControl.LoadDataContext() in C:\Users\kenne\Documents\ANTLR4ParseTreeVisualizer\Shared\VisualizerControl.xaml.cs:line 71
   at ParseTreeVisualizer.VisualizerControl.set_objectProvider(IVisualizerObjectProvider value) in C:\Users\kenne\Documents\ANTLR4ParseTreeVisualizer\Shared\VisualizerControl.xaml.cs:line 95
   at ParseTreeVisualizer.Visualizer.Show(IDialogVisualizerService windowService, IVisualizerObjectProvider objectProvider) in C:\Users\kenne\Documents\ANTLR4ParseTreeVisualizer\Shared\Visualizer.cs:line 42
   at Microsoft.VisualStudio.Debugger.ClrCustomVisualizerVSHostHelper.DisplayValue(IntPtr hWnd, SuccessEvaluationResultProxy evaluationResult, Int32 selectedIndex)
`

It took 30+ seconds to parse, which is a problem in itself. I'm looking into Harwell's optimized runtime to see if that will help.
